mono/utils/mono-threads.c

```diff
@@ -993,37 +1004,104 @@ mono_thread_info_resume (MonoNativeThreadId tid)
 	return result;
 }
 
-gboolean
-mono_thread_info_begin_suspend (MonoThreadInfo *info)
+static MonoThreadBeginSuspendResult
+begin_suspend_phase_coop (MonoThreadInfo *info);
+
+static MonoThreadBeginSuspendResult
+begin_suspend_phase_all (MonoThreadInfo *info);
+
+
+MonoThreadBeginSuspendResult
+mono_thread_info_begin_suspend (MonoThreadInfo *info, MonoThreadSuspendPhase phase)
+{
+	/* FIXME: rationalize this spaghetti */
+	if (phase == MONO_THREAD_SUSPEND_PHASE_ALL && mono_threads_is_hybrid_suspension_enabled ())
+		return begin_suspend_phase_all (info);
+	else
+		return begin_suspend_phase_coop (info);
+}
+
+MonoThreadBeginSuspendResult
+begin_suspend_phase_coop (MonoThreadInfo *info)
 {
 	switch (mono_threads_transition_request_suspension (info)) {
```

Comment:
@kumpera in full preemptive:
1. `mono_threads_transition_request_suspension` will only ever return `ReqSuspendInitSuspendRunning` or `ReqSuspendAlreadySuspended`.  So we will either call `begin_suspend_for_running_thread` or just return `MONO_THREAD_BEGIN_SUSPEND_SUSPENDED`.
2. `begin_suspend_for_running_thread` will call `begin_preemptive_suspend`
3. `begin_preemptive_suspend`  calls `mono_threads_suspend_begin_async_suspend` which starts the platform-specific async suspend.

Reply:
So back out in `sgen_unified_suspend_stop_world`: we start in the first phase, we set `need_next_phase = FALSE`  and we try to preemptively suspend each thread and get back either `MONO_THREAD_BEGIN_SUSPEND_SUSPENDED` or `MONO_THREAD_BEGIN_SUSPEND_SKIP` - never `MONO_THREAD_BEGIN_SUSPEND_NEXT_PHASE`.  So `need_next_phase` is never TRUE and we skip the mop-up phase.

Reply:
I'm going to close this PR soon - now that I can see what the hybrid phases are supposed to be doing, I think I can clean this up so that each suspend policy is more separated and it's easier to follow.